### PR TITLE
Cleanup 3

### PR DESCRIPTION
**Pull request by small-turtle-1**
### What problem does this PR solve?

1. Add cleanup of index, deprecate segment, deprecate segment index.
2. Fix delta catalog init process. (get parent name in constructor)
3. Use crab latch in catalog. So that Meta is able to cleaned up.
4. Refactor catalog to remove duplicate code.
5. Fix: bad generate method of random file path.

Issue link:
#670
#635

### Type of change

- [x] Bug Fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [x] Refactoring
- [ ] Performance Improvement
- [x] Test cases
- [ ] Python SDK impacted, Need to update PyPI
- [ ] Other (please describe):

### CODE REVIEW

File: src/common/utility/random.cpp [hunk 1]
```diff
@@ -16,22 +16,42 @@ module;
...
     }
     return ret;
 }
+
+SharedPtr<String> DetermineRandomString(const String &parent_dir, const String &name) {
+    LocalFileSystem fs;
+    SharedPtr<String> result;
+    int cnt = 0;
+    do {
+        u32 seed = std::rand();
```
Comment (JinHai-CN) [inline src/common/utility/random.cpp:49]: Use timestamp as the seed?

File: src/storage/bg_task/compact_segments_task.cpp [hunk 2]
```diff
@@ -197,7 +197,12 @@ void CompactSegmentsTask::CreateNewIndex(BaseTableRef *new_table_ref) {
             // Table index entry isn't found
             RecoverableError(status);
         }
-        txn_->CreateIndexPrepare(table_index_entry, new_table_ref, false, false);
+        table_index_entry->CreateIndexPrepare(table_entry,
+                                              new_table_ref->block_index_.get(),
+                                              txn_,
+                                              false,  // prepare
```
Comment (JinHai-CN) [inline src/storage/bg_task/compact_segments_task.cpp:203]: too many flags, hard to understand

File: src/storage/buffer/buffer_manager.cpp [hunk 3]
```diff
@@ -42,11 +42,10 @@ BufferObj *BufferManager::Allocate(UniquePtr<FileWorker> file_worker) {
     auto buffer_obj = MakeUnique<BufferObj>(this, true, std::move(file_worker));
 
     auto res = buffer_obj.get();
-    std::unique_lock<std::shared_mutex> w_locker(rw_locker_);
+    std::unique_lock w_locker(rw_locker_);
     if (auto iter = buffer_map_.find(file_path); iter != buffer_map_.end()) {
-        UniquePtr<String> err_msg = MakeUnique<String>(fmt::format("BufferManager::Allocate: file {} already exists.", file_path.c_str()));
-        LOG_ERROR(*err_msg);
-        UnrecoverableError(*err_msg);
+        iter->second->Cleanup();
```
Comment (JinHai-CN) [inline src/storage/buffer/buffer_manager.cpp:47]: Should raise exception.

File: src/storage/buffer/file_worker/file_worker.cpp [hunk 4]
```diff
@@ -109,7 +109,6 @@ void FileWorker::CleanupFile() {
     String file_path = fmt::format("{}/{}", *file_dir_, *file_name_);
     if (!fs.Exists(file_path)) {
         return;
-        // UnrecoverableError(fmt::format("File {} not found.", file_path));
```
Comment (JinHai-CN) [inline src/storage/buffer/file_worker/file_worker.cpp:112]: Should raise exception.

File: src/storage/meta/catalog.cpp [hunk 5]
```diff
@@ -56,71 +56,25 @@ import segment_column_index_entry;
...
-            db_meta = db_iter2->second.get();
-        }
-        this->rw_locker().unlock();
-    }
-
+        return DBMeta::NewDBMeta(db_dir, MakeShared<String>(db_name));
+    };
     LOG_TRACE(fmt::format("Adding new database entry: {}", db_name));
-    return db_meta->CreateNewEntry(txn_id, begin_ts, txn_mgr, conflict_type);
+    auto [db_meta, r_lock] = this->db_meta_map_.GetMeta(db_name, std::move(init_db_meta), txn_id, begin_ts, txn_mgr);
```
Comment (JinHai-CN) [inline src/storage/meta/catalog.cpp:76]: Readable issue?

File: src/storage/buffer/buffer_manager.cpp (hunk 3 above)
Comment (small-turtle-1) [reply src/storage/buffer/buffer_manager.cpp:47]: The file path is unique in all current files, but may duplicate with old files.
When cleanup task PickCleanup a file and remove it from catalog, the cleanup task will release lock. The actual cleanup task may process in parallel with adding file. The latter may has the same name with cleanuped file and try to Allocate here. So Cleanup the BufferObj here directly.

File: src/storage/meta/meta_map.cppm [hunk 6]
```diff
@@ -14,51 +14,135 @@
...
+        }
+        r_lock.lock();
+    } else {
+        LOG_TRACE("Add new entry in existed meta_map");
+    }
+    return {iter->second.get(), std::move(r_lock)};
+}
+
+template <MetaConcept Meta>
+Tuple<Meta *, Status, std::shared_lock<std::shared_mutex>> MetaMap<Meta>::GetExistMeta(const String &name, ConflictType conflict_type) {
```
Comment (JinHai-CN) [inline src/storage/meta/meta_map.cppm:97]: Two implementation:
1. Return shared locker
2. Return non-shared locker

File: src/storage/meta/catalog.cpp [hunk 7]
```diff
@@ -571,12 +505,12 @@ void Catalog::LoadFromEntry(Catalog *catalog, const String &catalog_path, Buffer
                 auto check_point_row_count = add_block_entry_op->checkpoint_row_count();
 
                 auto *db_meta = catalog->db_meta_map().at(db_name).get();
-                auto [db_entry, db_status] = db_meta->GetEntryReplay(txn_id, begin_ts);
+                auto [db_entry, db_status] = db_meta->GetEntry(txn_id, begin_ts);
```
Comment (JinHai-CN) [inline src/storage/meta/catalog.cpp:508]: Can't use GetEntryReplay to replace GetEntry

File: src/storage/meta/entry/base_entry.cppm [hunk 8]
```diff
@@ -38,11 +38,15 @@ export enum class EntryType : i8 {
     kBlockColumn,
 };
 
+export enum class EntryStatus : i8 {
```
Comment (JinHai-CN) [inline src/storage/meta/entry/base_entry.cppm:41]: Remove it

File: src/storage/meta/entry/base_entry.cppm [hunk 9]
```diff
@@ -59,16 +63,57 @@ public:
...
     atomic_u64 txn_id_{0};
     TxnTimeStamp begin_ts_{0};
     atomic_u64 commit_ts_{UNCOMMIT_TS};
-    bool deleted_{false};
+    bool deleted_ = false;
+
+    const EntryType entry_type_{EntryType::kDummy};
+
+public:
+    bool IfNotCleanedup(std::function<void()> &&func) {
```
Comment (JinHai-CN) [inline src/storage/meta/entry/base_entry.cppm:77]: remove IfNotCleanedup and SetCleanuped

File: src/storage/meta/entry/base_entry.cppm [hunk 10]
```diff
@@ -59,16 +63,57 @@ public:
...
+    std::mutex mtx_;
+
+    EntryStatus status_{EntryStatus::kInit};
 };
 
-// Merge two reverse-ordered list inplace.
-export void MergeLists(List<SharedPtr<BaseEntry>> &list1, List<SharedPtr<BaseEntry>> &list2);
+export struct BaseMeta {
+public:
+    bool IfNotCleanedup(std::function<void()> &&func) {
```
Comment (JinHai-CN) [inline src/storage/meta/entry/base_entry.cppm:99]: remove IfNotCleanedup and SetCleanuped

File: src/storage/meta/entry/db_entry.cpp [hunk 11]
```diff
@@ -33,16 +33,15 @@ import status;
...
-DBEntry::DBEntry() : BaseEntry(EntryType::kDummy) {}
-
-DBEntry::DBEntry(const SharedPtr<String> &data_dir, const SharedPtr<String> &db_name, TransactionID txn_id, TxnTimeStamp begin_ts)
+DBEntry::DBEntry(bool is_delete, const SharedPtr<String> &data_dir, const SharedPtr<String> &db_name, TransactionID txn_id, TxnTimeStamp begin_ts)
     // "data_dir": "/tmp/infinity/data"
     // "db_entry_dir": "/tmp/infinity/data/db1/txn_6"
-    : BaseEntry(EntryType::kDatabase), db_entry_dir_(MakeShared<String>(fmt::format("{}/{}/txn_{}", *data_dir, *db_name, txn_id))),
-      db_name_(db_name) {
+    : BaseEntry(EntryType::kDatabase), db_entry_dir_(is_delete ? nullptr : DetermineDBDir(*data_dir, *db_name)), db_name_(db_name) {
+    this->deleted_ = is_delete;
```
Comment (JinHai-CN) [inline src/storage/meta/entry/db_entry.cpp:44]: why not in initialize_list?

File: src/storage/meta/entry/db_entry.cpp [hunk 12]
```diff
@@ -75,104 +75,41 @@ Tuple<TableEntry *, Status> DBEntry::CreateTable(TableEntryType table_entry_type
...
-    if (this->table_meta_map().find(table_collection_name) != this->table_meta_map().end()) {
-        table_meta = this->table_meta_map()[table_collection_name].get();
-    }
-    this->rw_locker().unlock_shared();
-
-    LOG_TRACE(fmt::format("Remove a table/collection entry: {}", table_collection_name));
-    table_meta->DeleteNewEntry(txn_id, txn_mgr);
+void DBEntry::RemoveTableEntry(const String &table_name, TransactionID txn_id) {
+    auto [table_meta, _1, _2] = table_meta_map_.GetExistMeta(table_name, ConflictType::kError); // FIXME(sys)
+    LOG_TRACE(fmt::format("Remove a table/collection entry: {}", table_name));
```
Comment (JinHai-CN) [inline src/storage/meta/entry/db_entry.cpp:111]: Log: Remove a table entry {}

File: src/storage/meta/entry/entry_list.cppm [hunk 13]
```diff
@@ -14,46 +14,327 @@
...
+    enum class FindResult : u8 {
+        kFound,
+        kNotFound,
+        kUncommitted,
+        kUncommittedDelete,
+        kConflict,
+    };
+
 public:
+    using Iterator = typename List<SharedPtr<Entry>>::iterator;
```
Comment (JinHai-CN) [inline src/storage/meta/entry/entry_list.cppm:51]: Remove the declaration

File: src/storage/meta/meta_map.cppm [hunk 14]
```diff
@@ -14,51 +14,135 @@
...
     std::unique_lock lock(rw_locker_);
+    for (auto &[name, meta] : meta_map_) {
+        func(meta.get());
+    }
+}
 
+template <MetaConcept Meta>
+void MetaMap<Meta>::PickCleanup(CleanupScanner *scanner) {
+    std::unique_lock w_lock(rw_locker_);
     for (auto iter = meta_map_.begin(); iter != meta_map_.end();) {
```
Comment (JinHai-CN) [inline src/storage/meta/meta_map.cppm:130]: 1. Collect all meta_map value ptr, and save it to vector
2. Loop the vector to pick entry to clean up.

File: src/storage/meta/table_index_meta.cpp [hunk 15]
```diff
@@ -46,232 +46,42 @@ UniquePtr<TableIndexMeta> TableIndexMeta::NewTableIndexMeta(TableEntry *table_en
     return table_index_meta;
 }
 
-Tuple<TableIndexEntry *, Status> TableIndexMeta::CreateTableIndexEntry(const SharedPtr<IndexDef> &index_def,
+Tuple<TableIndexEntry *, Status> TableIndexMeta::CreateTableIndexEntry(std::shared_lock<std::shared_mutex> r_lock,
```
Comment (JinHai-CN) [inline src/storage/meta/table_index_meta.cpp:49]: std::shared_lock<std::shared_mutex>&&

File: src/storage/txn/txn.cpp [hunk 16]
```diff
@@ -169,7 +169,7 @@ Status Txn::CreateDatabase(const String &db_name, ConflictType conflict_type) {
     TxnTimeStamp begin_ts = txn_context_.GetBeginTS();
 
     auto [db_entry, status] = catalog_->CreateDatabase(db_name, this->txn_id_, begin_ts, txn_mgr_, conflict_type);
-    if (!status.ok()) {
+    if (db_entry == nullptr) {
```
Comment (JinHai-CN) [inline src/storage/txn/txn.cpp:172]: Add a comment to illustrate

File: src/storage/txn/txn.cpp [hunk 17]
```diff
@@ -309,7 +303,7 @@ Tuple<TableIndexEntry *, Status> Txn::CreateIndexDef(TableEntry *table_entry, co
     TxnTimeStamp begin_ts = txn_context_.GetBeginTS();
 
     auto [table_index_entry, index_status] = catalog_->CreateIndex(table_entry, index_def, conflict_type, txn_id_, begin_ts, txn_mgr_);
-    if (!index_status.ok() || (index_status.ok() && table_index_entry == nullptr && conflict_type == ConflictType::kIgnore)) {
+    if (table_index_entry == nullptr) {
```
Comment (JinHai-CN) [inline src/storage/txn/txn.cpp:306]: Add a comment

**Review by JinHai-CN — APPROVED** [review]

**Pull request merged**